File: Communication/UCI.cs
using System;
using System.Reflection;

namespace CHEPPP
{
    public class UCI
    {
        public UCI()
        {
            SendId();
            SendReady();
        }

        public void Start()
        {
            while (true)
            {
                string uciCommand = Console.ReadLine();
                if (uciCommand != null)
                {
                    if (ParseUciCommand(uciCommand.Trim().ToLower())) {
                        //command runs
                    }
                    else
                    {
                        //when false was sent break
                        Console.WriteLine("Exiting.");
                        break;
                    }
                }
            }
        }

        private bool ParseUciCommand(string uciCommand)
        {
            if (uciCommand == null)
            {
                Console.WriteLine("Invalid command");
                return true; // Invalid
            }
            string[] commandParts = uciCommand.Split(" ", StringSplitOptions.RemoveEmptyEntries);
            if (commandParts.Length == 0)
            {
                Console.WriteLine("Invalid command");
                return true; // Invalid
            }

            uciCommand = commandParts[0];

            switch (uciCommand)
            {
                case "quit":
                    return false;    
                case "isready":
                    IsReady();
                    break;
                case "position":
                    Position(commandParts);
                    break;
                default:
                    return true;
            }
            return true;
        }

        private void Position(string[] parts)
        {
            if (parts.Length < 2)
            {
                Console.WriteLine("Invalid command, Fen missing?");
                return; // Invalid
            }
        }

        private void SendId()
        {
            Console.WriteLine("id name CHEPPP 0.1"); // TODO: get version from assembly
            Console.WriteLine("id author Phil Baum");
        }

        private static void SendReady()
        {
            Console.WriteLine("uciok");
        }

        private void IsReady()
        {
            Console.WriteLine("readyok"); //TODO: Always ready?
        }
    }
}

File: UCI.py
import engine


class UCI:

    def __init__(self):
        self.send_id()
        self.send_ready()
        pass

    def send_id(self):
        print("id name Spot 0.1")
        print("id author Phil Baum")

    def send_ready(self):
        print("uciok")

    def is_ready(self):
        print("readyok")

    def validate_uci_command(self, uci_command):
        # TODO
        return True

    def perft(self, depth):
        e = engine.Engine()
        e.start_perft(depth)

    def parse_uci_command(self, uci_command):
        cmd_parts = uci_command.split()

        if cmd_parts[0] == "quit":
            return False
        elif cmd_parts[0] == "isready":
            self.is_ready()
        elif cmd_parts[0] == "uci":
            self.is_ready()
        elif cmd_parts[0] == "position":
            pass # TODO
        elif cmd_parts[0] == "go":
            pass # TODO
        elif cmd_parts[0] == "debug":
            pass # TODO
        elif cmd_parts[0] == "perft":
            self.perft(int(cmd_parts[1]))

        return True

    def start(self):

        while True:

            uci_command = input().strip()

            if self.validate_uci_command(uci_command):
                if not self.parse_uci_command(uci_command):
                    print("Exiting.")
                    break
